In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('data/wine_cleared.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             129971 non-null  int64  
 1   country                129908 non-null  object 
 2   description            129971 non-null  object 
 3   designation            129971 non-null  object 
 4   points                 129971 non-null  int64  
 5   price                  129971 non-null  float64
 6   province               129908 non-null  object 
 7   region_1               129971 non-null  object 
 8   taster_name            129971 non-null  object 
 9   taster_twitter_handle  129971 non-null  object 
 10  title                  129971 non-null  object 
 11  variety                129970 non-null  object 
 12  winery                 129971 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 12.9+ MB


In [4]:
df.drop(columns=['Unnamed: 0'], inplace=True)

In [5]:
df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  
0              Nicosia  
1  Quinta dos Avidagos  
2            Rainstorm  
3           St. Julian  
4         Sweet Cheeks

In [6]:
df['price_round'] = df['price'].round().astype('int')

In [7]:
df['year'] = df['title'].str.findall('\d{4}').str.get(0).astype('int', copy=False, errors='ignore')

In [8]:
df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  year  
0              Nicosia           35  2013  
1  Quinta dos Avidagos           15  2011  
2            Rainstorm           14  2013  
3           St. Julian           13  2013  
4         Sweet Cheeks           65  2012

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 14 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   country                129908 non-null  object 
 1   description            129971 non-null  object 
 2   designation            129971 non-null  object 
 3   points                 129971 non-null  int64  
 4   price                  129971 non-null  float64
 5   province               129908 non-null  object 
 6   region_1               129971 non-null  object 
 7   taster_name            129971 non-null  object 
 8   taster_twitter_handle  129971 non-null  object 
 9   title                  129971 non-null  object 
 10  variety                129970 non-null  object 
 11  winery                 129971 non-null  object 
 12  price_round            129971 non-null  int32  
 13  year                   125362 non-null  object 
dtypes: float64(1), int32(1), int64(1), o

In [10]:
df['is_usa'] = df['country'].apply(lambda x: 1 if x == 'US' else 0)

###  Задание 2.1 ###

Выбрать из списка две самых популярных (помимо США) страны, производящих вино.

***Ответ - France, Italy***

In [11]:
df['country'].value_counts()

US                        54504
France                    22093
Italy                     19540
Spain                      6645
Portugal                   5691
Chile                      4472
Argentina                  3800
Austria                    3345
Australia                  2329
Germany                    2165
New Zealand                1419
South Africa               1401
Israel                      505
Greece                      466
Canada                      257
Hungary                     146
Bulgaria                    141
Romania                     120
Uruguay                     109
Turkey                       90
Slovenia                     87
Georgia                      86
England                      74
Croatia                      73
Mexico                       70
Moldova                      59
Brazil                       52
Lebanon                      35
Morocco                      28
Peru                         16
Ukraine                      14
Serbia  

###  Задание 2.2 ###

Создать бинарные признаки `is_france`, `is_italy` наподобие признака `is_usa`.

В ответ вписать результат выполнения кода `data['is_france'].sum()` и `data['is_italy'].sum()`

***Ответ - 22093, 19540***

In [12]:
df['is_france'] = df['country'].apply(lambda x: 1 if x == 'France' else 0)
df['is_italy'] = df['country'].apply(lambda x: 1 if x == 'Italy' else 0)
df['is_france'].sum()

22093

In [13]:
df['is_italy'].sum()

19540

###  Задание 2.3 ###

Создать новый бинарный признак `old_wine`, где значение **1** будет обозначать, что вино старше 2010 года.

В ответ вписать результат выполнения кода `data['old_wine'].sum()`

***Ответ - 39781***

In [14]:
df['old_wine'] = df['year'].apply(lambda x: 1 if pd.to_numeric(x) < 2010 else 0)

In [15]:
df['old_wine'].sum()

39781

###  Задание 2.7 (Самопроверка) ###

Создать новый признак `locality` из признака `title`, который будет обозначать название долины/местности производства вина.

Например, в названии вина **Rainstorm 2013 Pinot Gris (Willamette Valley)** `locality` будет **Willamette Valley**. В названии **Tandem 2011 Ars In Vitro Tempranillo-Merlot (Navarra)**&nbsp;&mdash; **Navarra**.

In [16]:
df['locality'] = df['title'].str.findall(r'\(.+\)').str.get(0).str.replace('\(|\)', '', regex=True)

In [18]:
df.head()

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points      price           province  \
0                        Vulkà Bianco      87  35.363389  Sicily & Sardinia   
1                            Avidagos      87  15.000000              Douro   
2                             unknown      87  14.000000             Oregon   
3                Reserve Late Harvest      87  13.000000           Michigan   
4  Vintner's Reserve Wild Child Block      87  65.000000             Oregon   

              region_1         taster_name taster_twitter_handle  \
0                 Etna       Kerin O’Keefe          @kerinokeefe   
1              unknown          Roger Voss            @vossroger   
2    Willamette Valley        Paul Gregutt           @paulgwine    
3  Lake Michigan Shore  Alexander Peartree               unknown   
4    Willamette Valley        Paul Gregutt           @paulgwine    

                                               title         variety  \
0                  Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1      Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2      Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3  St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4  Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   

                winery  price_round  year  is_usa  is_france  is_italy  \
0              Nicosia           35  2013       0          0         1   
1  Quinta dos Avidagos           15  2011       0          0         0   
2            Rainstorm           14  2013       1          0         0   
3           St. Julian           13  2013       1          0         0   
4         Sweet Cheeks           65  2012       1          0         0   

   old_wine             locality  
0         0                 Etna  
1         0                Douro  
2         0    Willamette Valley  
3         0  Lake Michigan Shore  
4         0    Willamette Valley

# 3. Создание признаков. Внешние источники данных #

In [19]:
country_population = pd.read_csv('data/country_population.csv', sep=';')
display(country_population)

country     population
0               China  1,411,778,724
1               India  1,386,584,581
2                  US    333,022,386
3           Indonesia    271,350,000
4            Pakistan    225,200,000
..                ...            ...
236              Niue          1,549
237           Tokelau          1,501
238      Vatican City            825
239     Cocos Islands            573
240  Pitcairn Islands             40

[241 rows x 2 columns]

###  Задание 3.1 ###

Каково население Италии согласно датасету `country_population`? Ответ без пробелов и запятых.

***Ответ - 59097904***

In [20]:
display(country_population[country_population.country == 'Italy'].population)

24    59,097,904
Name: population, dtype: object

In [21]:
df.join(country_population.set_index('country'), on='country')

country                                        description  \
0          Italy  Aromas include tropical fruit, broom, brimston...   
1       Portugal  This is ripe and fruity, a wine that is smooth...   
2             US  Tart and snappy, the flavors of lime flesh and...   
3             US  Pineapple rind, lemon pith and orange blossom ...   
4             US  Much like the regular bottling from 2012, this...   
...          ...                                                ...   
129966   Germany  Notes of honeysuckle and cantaloupe sweeten th...   
129967        US  Citation is given as much as a decade of bottl...   
129968    France  Well-drained gravel soil gives this wine its c...   
129969    France  A dry style of Pinot Gris, this is crisp with ...   
129970    France  Big, rich and off-dry, this is powered by inte...   

                                   designation  points      price  \
0                                 Vulkà Bianco      87  35.363389   
1                                     Avidagos      87  15.000000   
2                                      unknown      87  14.000000   
3                         Reserve Late Harvest      87  13.000000   
4           Vintner's Reserve Wild Child Block      87  65.000000   
...                                        ...     ...        ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90  28.000000   
129967                                 unknown      90  75.000000   
129968                                   Kritt      90  30.000000   
129969                                 unknown      90  32.000000   
129970           Lieu-dit Harth Cuvée Caroline      90  21.000000   

                 province             region_1         taster_name  \
0       Sicily & Sardinia                 Etna       Kerin O’Keefe   
1                   Douro              unknown          Roger Voss   
2                  Oregon    Willamette Valley        Paul Gregutt   
3                Michigan  Lake Michigan Shore  Alexander Peartree   
4                  Oregon    Willamette Valley        Paul Gregutt   
...                   ...                  ...                 ...   
129966              Mosel              unknown  Anna Lee C. Iijima   
129967             Oregon               Oregon        Paul Gregutt   
129968             Alsace               Alsace          Roger Voss   
129969             Alsace               Alsace          Roger Voss   
129970             Alsace               Alsace          Roger Voss   

       taster_twitter_handle  \
0               @kerinokeefe   
1                 @vossroger   
2                @paulgwine    
3                    unknown   
4                @paulgwine    
...                      ...   
129966               unknown   
129967           @paulgwine    
129968            @vossroger   
129969            @vossroger   
129970            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 Pinot Noir (Oregon)      Pinot Noir   
129968  Domaine Gresser 2013 Kritt Gewurztraminer (Als...  Gewürztraminer   
129969      Domaine Marcel Deiss 2012 Pinot Gris (Alsace)      Pinot Gris   
129970  Domaine Schoffit 2012 Lieu-dit Harth Cuvée Car...  Gewürztraminer   

                                          winery  price_round  year  is_usa  \
0                                        Nicosia           35  2013       0   
1                 

In [22]:
country_area = pd.read_csv('data/country_area.csv', sep=';')
display(country_area)

country         area
0          Russia  17075400.00
1          Canada   9984670.00
2           China   9596960.00
3              US   9372610.00
4          Brazil   8547000.00
..            ...          ...
188    San Marino        61.00
189        Tuvalu        26.00
190         Nauru        21.00
191        Monaco         2.00
192  Vatican City         0.44

[193 rows x 2 columns]

###  Задание 3.2 ###

Создать новый признак `area_country`&nbsp;&mdash; площадь страны, аналогичный признаку `country_population`.

Какая площадь страны у вина под названием **'Gård 2014 Grand Klasse Reserve Lawrence Vineyards Viognier (Columbia Valley (WA))'**?

Ответ без пробелов, округлить до целых.

***Ответ - 9372610***

In [23]:
df2 = df.join(country_area.set_index('country'), on='country')

In [24]:
df2[df2['title'].str.contains('Reserve Lawrence Vineyards Viognier')].area

94    9372610.0
Name: area, dtype: float64

# 4. Создание признаков. Работа с форматом &laquo;дата-время&raquo; #

In [25]:
calls_list = [
    [460, '2013-12-17 04:55:39', '2013-12-17 04:55:44', '2013-12-17 04:55:45'],
    [12, '2013-12-16 20:03:20', '2013-12-16 20:03:22', '2013-12-16 20:07:13'],
    [56, '2013-12-16 20:03:20', '2013-12-16 20:03:20', '2013-12-16 20:05:04'],
    [980, '2013-12-16 20:03:20','2013-12-16 20:03:27', '2013-12-16 20:03:29'],
    [396, '2013-12-16 20:08:27', '2013-12-16 20:08:28','2013-12-16 20:12:03'],
    [449, '2013-12-16 20:03:20', '2013-12-16 20:03:25','2013-12-16 20:05:00'],
    [397, '2013-12-16 20:08:25', '2013-12-16 20:08:27', '2013-12-16 20:09:59'],
    [398, '2013-12-16 20:01:23', '2013-12-16 20:01:23', '2013-12-16 20:04:58'],
    [452, '2013-12-16 20:03:20', '2013-12-16 20:03:21','2013-12-16 20:04:55'],
    [440, '2013-12-16 20:03:20', '2013-12-16 20:04:26', '2013-12-16 20:04:32']
]
calls = pd.DataFrame(calls_list, columns=['client_id', 'agent_date', 'created_at', 'end_date'])
calls['agent_date'] = pd.to_datetime(calls['agent_date'])
calls['created_at'] = pd.to_datetime(calls['created_at'])
calls['end_date'] = pd.to_datetime(calls['end_date'])
display(calls)

client_id          agent_date          created_at            end_date
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32

In [26]:
calls.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   client_id   10 non-null     int64         
 1   agent_date  10 non-null     datetime64[ns]
 2   created_at  10 non-null     datetime64[ns]
 3   end_date    10 non-null     datetime64[ns]
dtypes: datetime64[ns](3), int64(1)
memory usage: 448.0 bytes


In [27]:
calls['duration'] = (calls['end_date'] - calls['created_at']).dt.seconds
display(calls)

client_id          agent_date          created_at            end_date  \
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45   
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13   
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04   
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29   
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03   
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00   
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59   
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58   
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55   
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32   

   duration  
0         1  
1       231  
2       104  
3         2  
4       215  
5        95  
6        92  
7       215  
8        94  
9         6

###  Задание 4.1 ###

Подсчитать, сколько секунд тратят сотрудники компании на дозвон клиенту. Результат записать в новый признак `time_connection`.

В ответ записать результат выполнения следующего кода: `calls['time_connection'].sum()`.

***Ответ - 89***

In [28]:
calls['time_connection'] = (calls['created_at'] - calls['agent_date']).dt.seconds
calls['time_connection'].sum()

89

###  Задание 4.2 ###

Создать новый признак `is_connection`&nbsp;&mdash; факт соединения с клиентом. Признак будет равен 1 в случае, если разговор состоялся и продлился больше 10 секунд, иначе&nbsp;&mdash; 0.

В ответ записать результат выполнения следующего кода: `calls['is_connection'].sum()`.

***Ответ - 7***

In [29]:
calls['is_connection'] = calls['duration'].apply(lambda x: 1 if x > 10 else 0)
calls['is_connection'].sum()

7

###  Задание 4.3 ###

Создать признак `time_diff`&nbsp;&mdash; разницу в секундах между началом звонка(не разговора) и его окончанием.

В ответ записать результат выполнения следующего кода: `calls['time_diff'].sum()`.

***Ответ - 1144***

In [30]:
calls['time_diff'] = (calls['end_date'] - calls['agent_date']).dt.seconds
calls['time_diff'].sum()

1144

In [31]:
display(calls)

client_id          agent_date          created_at            end_date  \
0        460 2013-12-17 04:55:39 2013-12-17 04:55:44 2013-12-17 04:55:45   
1         12 2013-12-16 20:03:20 2013-12-16 20:03:22 2013-12-16 20:07:13   
2         56 2013-12-16 20:03:20 2013-12-16 20:03:20 2013-12-16 20:05:04   
3        980 2013-12-16 20:03:20 2013-12-16 20:03:27 2013-12-16 20:03:29   
4        396 2013-12-16 20:08:27 2013-12-16 20:08:28 2013-12-16 20:12:03   
5        449 2013-12-16 20:03:20 2013-12-16 20:03:25 2013-12-16 20:05:00   
6        397 2013-12-16 20:08:25 2013-12-16 20:08:27 2013-12-16 20:09:59   
7        398 2013-12-16 20:01:23 2013-12-16 20:01:23 2013-12-16 20:04:58   
8        452 2013-12-16 20:03:20 2013-12-16 20:03:21 2013-12-16 20:04:55   
9        440 2013-12-16 20:03:20 2013-12-16 20:04:26 2013-12-16 20:04:32   

   duration  time_connection  is_connection  time_diff  
0         1                5              0          6  
1       231                2              1        233  
2       104                0              1        104  
3         2                7              0          9  
4       215                1              1        216  
5        95                5              1        100  
6        92                2              1         94  
7       215                0              1        215  
8        94                1              1         95  
9         6               66              0         72

In [32]:
calls.drop(columns=['agent_date', 'created_at', 'end_date'], inplace=True)
display(calls)

client_id  duration  time_connection  is_connection  time_diff
0        460         1                5              0          6
1         12       231                2              1        233
2         56       104                0              1        104
3        980         2                7              0          9
4        396       215                1              1        216
5        449        95                5              1        100
6        397        92                2              1         94
7        398       215                0              1        215
8        452        94                1              1         95
9        440         6               66              0         72

###  Задание 4.5 ###

Создать признак количество дней с момента произведения вина&nbsp;&mdash; `years_diff` для датасета винных обзоров. За дату отсчёта взять **12 января 2022 года**. В ответ вписать максимальное количество дней с момента произведения вина.

Ответ округлить до целого числа.

***Ответ - 98627***

In [171]:
dfc = df.copy()
dfc.drop(columns=['country', 'province', 'variety', 'locality', 'is_usa', 'is_france', 'is_italy'], inplace=True)
dfc.dropna(inplace=True)
list_ini = ['1000', '1070', '1150', '1492', '1503', '1607', '1621', '1637', '1752', '1789',
            '1821', '1827', '1840', '1845', '1847', '1848', '1850', '1852', '1856', '1860',
            '1868', '1870', '1872', '1875', '1877', '1882', '1887', '1898', '1904', '1912',
            '1919', '1927', '1929', '1934', '1935', '1941', '1945', '1947', '1952', '1957',
            '1961', '1963', '1964', '1965', '1966', '1967', '1968', '1969', '1973', '1974',
            '1976', '1978', '1980', '1982', '1983', '1984', '1985', '1986', '1987', '1988',
            '1989', '1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997', '1998',
            '1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007', '2008',
            '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '3000',
            '7200']
dfc.drop(index=dfc[dfc.year < '1752'].index, inplace=True)
dfc.drop(index=dfc[dfc.year == '3000'].index, inplace=True)
dfc.drop(index=dfc[dfc.year == '7200'].index, inplace=True)
dfc['year'].unique()

array(['2013', '2011', '2012', '2010', '2007', '2009', '2008', '2014',
       '2015', '2016', '2004', '2003', '2006', '2001', '2005', '2002',
       '1887', '2000', '1999', '1991', '1997', '1996', '1877', '2017',
       '1995', '1872', '1868', '1898', '1998', '1852', '1994', '1992',
       '1840', '1929', '1912', '1875', '1976', '1964', '1848', '1870',
       '1856', '1983', '1967', '1990', '1988', '1827', '1860', '1850',
       '1980', '1987', '1989', '1993', '1969', '1882', '1935', '1821',
       '1973', '1978', '1965', '1968', '1947', '1963', '1985', '1927',
       '1904', '1847', '1982', '1986', '1752', '1789', '1919', '1957',
       '1966', '1984', '1961', '1845', '1952', '1941', '1974', '1934',
       '1945'], dtype=object)

In [172]:
# dfc[dfc.year == '1752']
dfc['days_old'] = pd.to_datetime('2022-01-12') - pd.to_datetime(dfc['year'])
dfc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 125331 entries, 0 to 129970
Data columns (total 13 columns):
 #   Column                 Non-Null Count   Dtype          
---  ------                 --------------   -----          
 0   description            125331 non-null  object         
 1   designation            125331 non-null  object         
 2   points                 125331 non-null  int64          
 3   price                  125331 non-null  float64        
 4   region_1               125331 non-null  object         
 5   taster_name            125331 non-null  object         
 6   taster_twitter_handle  125331 non-null  object         
 7   title                  125331 non-null  object         
 8   winery                 125331 non-null  object         
 9   price_round            125331 non-null  int32          
 10  year                   125331 non-null  object         
 11  old_wine               125331 non-null  int64          
 12  days_old               125331 

In [173]:
dfc['days_old'].max()

Timedelta('98627 days 00:00:00')

# 5. Кодирование признаков. Методы #

In [188]:
##### pip install category_encoders
import category_encoders as ce

In [189]:
clothing_list = [
    ['xxs', 'dress'],
    ['xxs', 'skirt'],
    ['xs', 'dress'],
    ['s', 'skirt'],
    ['m', 'dress'],
    ['l', 'shirt'],
    ['s', 'coat'],
    ['m', 'coat'],
    ['xxl', 'shirt'],
    ['l', 'dress']
]
clothing = pd.DataFrame(clothing_list, columns=['size', 'type'])
display(clothing)

size   type
0  xxs  dress
1  xxs  skirt
2   xs  dress
3    s  skirt
4    m  dress
5    l  shirt
6    s   coat
7    m   coat
8  xxl  shirt
9    l  dress

In [190]:
ord_encoder = ce.OrdinalEncoder()
data_bin = ord_encoder.fit_transform(clothing[['size', 'type']])
clothing2 = pd.concat([clothing, data_bin], axis=1)
display(clothing2)

size   type  size  type
0   xxs  dress     1     1
1   xxs  skirt     1     2
2    xs  dress     2     1
3     s  skirt     3     2
4     m  dress     4     1
5     l  shirt     5     3
6     s   coat     3     4
7     m   coat     4     4
8   xxl  shirt     6     3
9     l  dress     5     1

###  Задание 5.3 (Самопроверка) ###

Использовать ранее изученные методы кодирования и закодировать признак `year` в датасете винных обзоров порядковым кодированием.

In [191]:
data_bin = ord_encoder.fit_transform(df[['year']])
display(data_bin)

year
0          1
1          2
2          1
3          1
4          3
...      ...
129966     1
129967    12
129968     1
129969     3
129970     3

[129971 rows x 1 columns]

## Однократное кодирование. One-Hot Encoding ##

In [192]:
encoder = ce.OneHotEncoder(cols=['type'])
type_bin = encoder.fit_transform(clothing['type'])
clothing3 = pd.concat([clothing, type_bin], axis=1)
display(clothing3)

size   type  type_1  type_2  type_3  type_4
0  xxs  dress       1       0       0       0
1  xxs  skirt       0       1       0       0
2   xs  dress       1       0       0       0
3    s  skirt       0       1       0       0
4    m  dress       1       0       0       0
5    l  shirt       0       0       1       0
6    s   coat       0       0       0       1
7    m   coat       0       0       0       1
8  xxl  shirt       0       0       1       0
9    l  dress       1       0       0       0

###  Задание 5.4 ###

В наборе данных винных обзоров признак, обозначающий имя сомелье (`taster_name`), является номинальным. Закодировать его, используя **One-Hot Encoding**.

В ответе написать, сколько признаков добавилось после применения кодирования.

***Ответ - 20***

In [195]:
encoder2 = ce.OneHotEncoder(cols=['taster_name'])
taster_hot = encoder2.fit_transform(df['taster_name'])
taster_hot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 20 columns):
 #   Column          Non-Null Count   Dtype
---  ------          --------------   -----
 0   taster_name_1   129971 non-null  int64
 1   taster_name_2   129971 non-null  int64
 2   taster_name_3   129971 non-null  int64
 3   taster_name_4   129971 non-null  int64
 4   taster_name_5   129971 non-null  int64
 5   taster_name_6   129971 non-null  int64
 6   taster_name_7   129971 non-null  int64
 7   taster_name_8   129971 non-null  int64
 8   taster_name_9   129971 non-null  int64
 9   taster_name_10  129971 non-null  int64
 10  taster_name_11  129971 non-null  int64
 11  taster_name_12  129971 non-null  int64
 12  taster_name_13  129971 non-null  int64
 13  taster_name_14  129971 non-null  int64
 14  taster_name_15  129971 non-null  int64
 15  taster_name_16  129971 non-null  int64
 16  taster_name_17  129971 non-null  int64
 17  taster_name_18  129971 non-null  int64
 18  tast

## Двоичное кодирование ##

In [197]:
bin_encoder = ce.BinaryEncoder(cols=['type'])
type_bin = bin_encoder.fit_transform(clothing['type'])
clothing4 = pd.concat([clothing, type_bin], axis=1)
display(clothing4)

size   type  type_0  type_1  type_2
0  xxs  dress       0       0       1
1  xxs  skirt       0       1       0
2   xs  dress       0       0       1
3    s  skirt       0       1       0
4    m  dress       0       0       1
5    l  shirt       0       1       1
6    s   coat       1       0       0
7    m   coat       1       0       0
8  xxl  shirt       0       1       1
9    l  dress       0       0       1

###  Задание 5.5 ###

Закодировать признак `country` двоичным способом.

Сколько новых признаков образовалось после кодирования признака country?

***Ответ - 6***

In [199]:
bin_encoder2 = ce.BinaryEncoder(cols=['country'])
type_bin2 = bin_encoder2.fit_transform(df['country'])
type_bin2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 6 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   country_0  129971 non-null  int64
 1   country_1  129971 non-null  int64
 2   country_2  129971 non-null  int64
 3   country_3  129971 non-null  int64
 4   country_4  129971 non-null  int64
 5   country_5  129971 non-null  int64
dtypes: int64(6)
memory usage: 5.9 MB


###  Задание 5.8 ###

На основе изученного материала определить подходящий способ кодирования признака `taster_twitter_handle` из датасета винных обзоров и закодировать его.

В ответе написать, сколько признаков добавилось после применения кодирования.

***Ответ - 5***

In [200]:
encoder3w = ce.OneHotEncoder(cols=['taster_twitter_handle'])
taster_tw = encoder3w.fit_transform(df['taster_twitter_handle'])
taster_tw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 16 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   taster_twitter_handle_1   129971 non-null  int64
 1   taster_twitter_handle_2   129971 non-null  int64
 2   taster_twitter_handle_3   129971 non-null  int64
 3   taster_twitter_handle_4   129971 non-null  int64
 4   taster_twitter_handle_5   129971 non-null  int64
 5   taster_twitter_handle_6   129971 non-null  int64
 6   taster_twitter_handle_7   129971 non-null  int64
 7   taster_twitter_handle_8   129971 non-null  int64
 8   taster_twitter_handle_9   129971 non-null  int64
 9   taster_twitter_handle_10  129971 non-null  int64
 10  taster_twitter_handle_11  129971 non-null  int64
 11  taster_twitter_handle_12  129971 non-null  int64
 12  taster_twitter_handle_13  129971 non-null  int64
 13  taster_twitter_handle_14  129971 non-null  int64
 14  taster_twitter_handl

In [202]:
encoder4w = ce.BinaryEncoder(cols=['taster_twitter_handle'])
type_bin2w = encoder4w.fit_transform(df['taster_twitter_handle'])
type_bin2w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 5 columns):
 #   Column                   Non-Null Count   Dtype
---  ------                   --------------   -----
 0   taster_twitter_handle_0  129971 non-null  int64
 1   taster_twitter_handle_1  129971 non-null  int64
 2   taster_twitter_handle_2  129971 non-null  int64
 3   taster_twitter_handle_3  129971 non-null  int64
 4   taster_twitter_handle_4  129971 non-null  int64
dtypes: int64(5)
memory usage: 5.0 MB


###  Задание 5.9 (Самопроверка) ###

Используйте следующий датафрейм для задания.

Определить типы признаков и закодировать их в соответствии с изученными способами.

In [204]:
list_of_dicts = [
 {'product': 'Product1', 'price': 1200, 'payment_type': 'Mastercard'},
 {'product': 'Product2', 'price': 3600, 'payment_type': 'Visa'},
 {'product': 'Product3', 'price': 7500, 'payment_type': 'Amex'}
]
dfs = pd.DataFrame(list_of_dicts)
display(dfs)

product  price payment_type
0  Product1   1200   Mastercard
1  Product2   3600         Visa
2  Product3   7500         Amex

In [208]:
# encoder_b = ce.BinaryEncoder(cols=['product'])
# type_b = encoder_b.fit_transform(dfs['product'])
type_b = ce.BinaryEncoder(cols=['product']).fit_transform(dfs['product'])
type_b.info()
type_h = ce.OneHotEncoder(cols=['payment_type']).fit_transform(dfs['payment_type'])
type_h.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   product_0  3 non-null      int64
 1   product_1  3 non-null      int64
dtypes: int64(2)
memory usage: 176.0 bytes
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   payment_type_1  3 non-null      int64
 1   payment_type_2  3 non-null      int64
 2   payment_type_3  3 non-null      int64
dtypes: int64(3)
memory usage: 200.0 bytes


# 6. Преобразование признаков. Нормализация. Стандартизация #